html_validation_fix

peter1000 opened the pull request:
Fix for `html` not passing validation issue #65.

This does not address all possible cases and may fail in certain situations.
But it seems to me that it is still a good improvement over the current situation.


http://validator.w3.org/#validate_by_input

**CURRENT  SITUATION**

Lexicon html api:

    Errors found while checking this document as HTML5!
    Result: 	10 Errors, 2 warning(s) 

Docile html api:

    Errors found while checking this document as HTML5!
    Result: 	56 Errors, 2 warning(s) 

Docile.Interface html api:

    Errors found while checking this document as HTML5!
    Result: 	10 Errors, 2 warning(s) 

**AFTER FIX**

Lexicon html api:

    This document was successfully checked as HTML5!
    Result: 	Passed, 2 warning(s)

Docile html api:

    This document was successfully checked as HTML5!
    Result: 	Passed, 2 warning(s)

Docile.Interface html api:

    This document was successfully checked as HTML5!
    Result: 	Passed, 2 warning(s)
    
---

* This does not bother to produce nice id's. It is just a simple fix.
    http://www.w3.org/TR/html4/types.html#type-id
    
    ID and NAME tokens must begin with a letter ([A-Za-z]) and may be followed by any number of letters, 
    digits ([0-9]), hyphens ("-"), underscores ("_"), colons (":"), and periods (".").
    
* This does not  check for duplicate id names in case of special character replacements.

* uses an ugly workaround to address potential `definition names` with unicode

```julia
function 但它不是一个大()
end
```

* This does not address other potential errors like `id` generated for definitions which only differ
in `case` like:

```julia
function a()
end

function A()
end
```

http://www.w3schools.com/tags/att_global_id.asp

`id 	Specifies a unique id for the element. Naming rules:`

* Must contain at least one character
* Must not contain any space characters
* **In HTML, all values are case-insensitive**


Other Changes: 

* by default typealias are also included in the output
* changed String => AbstractString

MichaelHatherly commented on src/render/html.jl:
```diff
@@ -153,8 +165,39 @@ function footerhtml(io::IO, doc::Metadata, config::Config)
+        else
+            id = id * string(Int(c))
+        end
+    end
```
Could use an `IOBuffer` here instead of `id = ""`, append strings to it using `write`, and then get the final result with `takebuf_string`. Should be faster for longer ids since it won't create a new string on every loop iteration.

peter1000 replied:
good point will change that

MichaelHatherly commented:
Other than the one comment, this is good to go.

Unfortunately, the tagged commit for version 0.1.6 didn't get pushed to github properly last night, which meant I've had to revert and reapply your compat.jl fixed, sorry about that. (You're still marked as the author by git though.) Could you rebase over the current master.

peter1000 commented:
>  I've had to revert and reapply your compat.jl fixed, sorry about that. (You're still marked as the author by git though.)

No problem at all. I will change the part you pointed out and rebase and update the PR.

peter1000 commented:
I checked `IOBuffer` is quite a bit faster and uses less memory. 

Just an unrelated **question**:

Are you already extracting somewhere in Docile: `kwarg` - or any suggestion how we could `easily` ;)  add it?

MichaelHatherly commented on src/render.jl:
```diff
@@ -56,17 +56,6 @@ function save(file::String, modulename::Module; args...)
-
-function push!(ents::Entries, modulename::Module, obj, ent::AbstractEntry)
-    push!(ents.entries, (modulename, obj, ent))
-end
```
@peter1000 what's the reasoning behind splitting `Entries` into `EntriesHtml` and `EntriesMD`? They look the same to me.

peter1000 commented:
> @peter1000 what's the reasoning behind splitting Entries into EntriesHtml and EntriesMD? They look the same to me.

Originally they where the same copy. But I changed the one for `markdown` and I got a conflict. So I renamed them for the upcoming other changes.

Anyway my plan for a later version would be to move it to the general render.jl and have later only one more generic one which could be useful for all formats.

I just wanted to start here somehow with a smaller PR and also fix the validation issue.

MichaelHatherly commented:
Ok, that's fine, just thought it was an unintentional change that slipped in :) Separating unrelated changes into separate commits helps sometimes when we need to go back through the commit history.
